http/handler.go

```diff
@@ -109,6 +110,24 @@ func (h *handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
+	// If we have a request body, make sure the preamble
+	// knows that it should close the body if it wants to
+	// write before completing reading.
+	// FIXME: https://github.com/ipfs/go-ipfs/issues/5168
+	// FIXME: https://github.com/golang/go/issues/15527
+	var bodyEOFChan chan struct{}
+	if r.Body != http.NoBody {
+		bodyEOFChan = make(chan struct{})
+		var once sync.Once
+		bw := bodyWrapper{
+			ReadCloser: r.Body,
+			onEOF: func() {
+				once.Do(func() { close(bodyEOFChan) })
+			},
+		}
+		r.Body = bw
+	}
```

Comment:
You may need to pre-close this channel in the else case.

Reply:
see https://github.com/ipfs/go-ipfs-cmds/pull/116#issuecomment-426586622